### ISSUE TITLE
include statement with wild card does not use a predictable order

### ISSUE THREAD

**Issue by Fjan** [ledger/ledger#1659]
I have a directory with the files Q1.ledger, Q2.ledger, Q3.ledger, Q4.ledger and a main ledger file that has the line: `include Q*.ledger`

However, ledger includes the files in seemingly random order, not Q1, Q2,… as hoped. This causes a problem if you have any expressions such as `check` statements that are dependent on the correct order. The order is consistent between runs, but not from one machine to another which caused a hard to track problem for us.

**How to reproduce**
Create a test.ledger file with the line `include Q*.ledger` and several files Q{1,2,3,4}.ledger with this:
```
2018-01-01 Test
    Assets                       1000.00 USD
    Test

check account("Assets").amount == 999.00 USD
```
Now a `ledger -f test.ledger bal`  will give:
```
Warning: "…/Q4.ledger", line 5: Check failed: account("Assets").amount == 999.00 USD
Warning: "…/Q2.ledger", line 5: Check failed: account("Assets").amount == 999.00 USD
Warning: "…/Q1.ledger", line 5: Check failed: account("Assets").amount == 999.00 USD
Warning: "…/Q3.ledger", line 5: Check failed: account("Assets").amount == 999.00 USD
```

**Mitigation**
The obvious workaround is to rewrite the `include Q*.ledger` to this:
```
include Q1.ledger
include Q2.ledger
include Q3.ledger
…
```
While the mitigation is easy enough, albeit a little cumbersome, it would be a nice feature if the include files were sorted to prevent anyone else running into unexpected behaviour.

**Comment by tbm** [comment]
I'm not sure it's entirely random since I always get the same result but I agree that sorting it would be best.

**Comment by scfc** [comment]
Could this be related/an artefact of issue #559 (and thus maybe fixed in `next`)?

**Comment by tbm** [comment]
@Fjan can you tested if this was fixed by dfd807f  as suggested by @scfc

**Comment by Fjan** [comment]
Hi @tbm, thanks for looking into this. I'm afraid I don't have an environment available to build ledger from source on a Mac (I always install through brew). If there is a beta version somewhere I can download I would be happy to give it a try. However, if the issue still exists it should be very easy to reproduce with the description above.

**Comment by tbm** [comment]
I just tried and the order is stable now, so this was fixed by dfd807f as suggested by @scfc

However, the order is not sorted (I get Q3, Q4, Q1, Q2 here, but reliably the same result).

**Comment by Fjan** [comment]
Hi @tbm. The order was always stable on one machine, but it changes from one machine to the next. The order can also change if you switch to a different branch if you have your ledger files in git.

In my opinion an unpredictable order is not useful behaviour for include, I would argue that the only order that makes sense is sorted. `include Q*.ledger` is useless if it doesn't include the quarters in order.

**Comment by tbm** [comment]
Ok, you're making a good point.  I agree.

When you originally reported this bug, the order changed between different runs of ledger.  This is no longer the case.  But as you say, it would still differ based on other factors.

**Comment by emin63** [comment]
Any update on this? I am a new ledger user and had a lot of trouble figuring out why my balance assertions were failing due to the use of the `include vanguard/*.dat` not reading files in a sane order. I discovered the order files were being read using the `--verbose` flag if that helps.

I think there are a couple of problems here:
  1. The documentation for the `include` directive is very terse and doesn't mention that the order files are read is strange.
  2. Not having a sane order makes using wildcards with include not particularly useful if files are time ordered.

If you prefer, I can open a new issue with a subject like "include statement with wild card parses in wrong order".

Thanks.

**Comment by tbm** [comment]
@emin63 there has been no progress but a PR would be very welcome.

> If you prefer, I can open a new issue with a subject like "include statement with wild card parses in wrong order".

I don't see how opening a new issue about the same issue is going to help?  What would the new issue cover that this one doesn't?

**Comment by emin63** [comment]
> @emin63 there has been no progress but a PR would be very welcome.

OK. I can take a look and see if it is something I am capable of fixing. Can you give me a pointer to the relevant parts of the source code I should look at? When I search for `include` I get a hit on basically every file because of the `#include`.

Thanks.

> I don't see how opening a new issue about the same issue is going to help? 

Agreed. There was some discussion earlier in the thread that this was closed because the order was stable. I didn't know if the problem was lack of time to work on this or lack of clarity on desired behavior.

Thanks.

**Comment by tbm** [comment]
instance_t::include_directive in textual.cc

**Comment by emin63** [comment]
> instance_t::include_directive in textual.cc

Thanks. See proposed fix in PR https://github.com/ledger/ledger/pull/1951

**Comment by jwiegley** [comment]
This is now merged, please confirm!

**Comment by tbm** [comment]
This was fixed in commit 8e55ba6929d1365e3f3f9441adcc94c92bc35330

**Comment by emin63** [comment]
Sorry for not confirming earlier. I didn't know if jwiegley was talking to me or tbm. In any case, I confirm the fix in master works for me. Thanks again for this great software.
